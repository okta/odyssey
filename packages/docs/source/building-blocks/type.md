# Type

Okta’s typography system is designed for clarity in readability and hierarchy.

## Hierarchy

Typographical hierarchy indicates importance of content. Through size and weight, a strong hierarchy can provide clarity to a user to easily understand which level of the content they’re on. Sections need subsections, and those subsections may also need their own subsections. Visually, a user should be able to clearly tell a parent section from a child section simply by the relationship in size and weight between headings.

### Scale

To ensure consistency in typography sizes throughout Okta touchpoints, we’ve created a custom modular type scale.

<figure class="ods-table--figure">
  <table class="ods-table type-sample--table">
    <thead>
      <tr>
        <th scope="column">
          rem
        </th>
        <th scope="column">
          px
        </th>
        <th scope="column">
          function
        </th>
        <th scope="column">
          example
        </th>
      </tr>
    </thead>
    <tbody>
      <tr class="type-sample">
        <td class="type-sample--rem"></td>
        <td class="type-sample--px"></td>
        <td class="type-sample--function"></td>
        <td class="type-sample--example">Waltz, bad nymph, for quick jigs vex!</td>
      </tr>
      <tr class="type-sample">
        <td class="type-sample--rem"></td>
        <td class="type-sample--px"></td>
        <td class="type-sample--function"></td>
        <td class="type-sample--example">Waltz, bad nymph, for quick jigs vex!</td>
      </tr>
      <tr class="type-sample">
        <td class="type-sample--rem"></td>
        <td class="type-sample--px"></td>
        <td class="type-sample--function"></td>
        <td class="type-sample--example">Waltz, bad nymph, for quick jigs vex!</td>
      </tr>
      <tr class="type-sample">
        <td class="type-sample--rem"></td>
        <td class="type-sample--px"></td>
        <td class="type-sample--function"></td>
        <td class="type-sample--example">Waltz, bad nymph, for quick jigs vex!</td>
      </tr>
      <tr class="type-sample">
        <td class="type-sample--rem"></td>
        <td class="type-sample--px"></td>
        <td class="type-sample--function"></td>
        <td class="type-sample--example">Waltz, bad nymph, for quick jigs vex!</td>
      </tr>
      <tr class="type-sample">
        <td class="type-sample--rem"></td>
        <td class="type-sample--px"></td>
        <td class="type-sample--function"></td>
        <td class="type-sample--example">Waltz, bad nymph, for quick jigs vex!</td>
      </tr>
      <tr class="type-sample">
        <td class="type-sample--rem"></td>
        <td class="type-sample--px"></td>
        <td class="type-sample--function"></td>
        <td class="type-sample--example">Waltz, bad nymph, for quick jigs vex!</td>
      </tr>
    </tbody>
  </table>
</figure>

Each step up in the modular scale is proportionally larger than the step before. The function looks like this:

```scss
@function ms($step) {
  $value: 1rem;

  @if $step > 0 {
    @for $i from 1 through $step {
      $value: ($value * $scale-ratio);
    }
  }

  @if $step < 0 {
    @for $i from $step through -1 {
      $value: ($value / $scale-ratio);
    }
  }

  @return $value;
}
```

### Weight

In order to maintain consistency across browsers and typefaces, we specify weights by their numbered value, not name. Currently, the following weights are available:

<figure class="ods-table--figure">
  <table class="ods-table">
    <thead>
      <tr>
        <th scope="column">
          Weight
        </th>
        <th scope="column">
          Example
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <code>font-weight: 400</code>
        </td>
        <td class="type-sample--400">
          Normal 400
        </td>
      </tr>
      <tr>
        <td>
          <code>font-weight: 600</code>
        </td>
        <td class="type-sample--600">
          Semi-bold 600
        </td>
      </tr>
    </tbody>
  </table>
</figure>

## Line-height

Our type variants include two <a href="https://developer.mozilla.org/en-US/docs/Web/CSS/line-height">line-heights</a> available for use:

<figure class="ods-table--figure">
  <table class="ods-table">
    <thead>
      <tr>
        <th scope="column">
          Line-height
        </th>
        <th scope="column"
          Variable
        </th>
        <th scope="column">
          Ratio
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Body Copy &amp; Standard UI
        </td>
        <td>
          <code>$base-line-height</code>
        </td>
        <td>
          1:1.5
        </td>
      </tr>
      <tr>
        <td>
          Headings
        </td>
        <td>
          <code>$heading-line-height</code>
        </td>
        <td>
          1:1.2
        </td>
      </tr>
    </tbody>
  </table>
</figure>

Our heading line-height should be used for any `font-size` set to `ms(4)` or higher.

### Accessibility Concerns

Our body copy line-height is based on the <a href="https://developer.mozilla.org/en-US/docs/Web/CSS/line-height#Accessibility_concerns">MDN recommendation</a>. A minimum value of 1.5 help text remain legible for low-vision users, as well as those with cognitive concerns.

Unitless values ensure that page zooming behaves as expected.

## Line Length

For inline and block level content, the line length should be 50 to 75 characters long. In order to facilitate consistency across designs, `max-width: $max-line-length` can be helpful.

## Heading Implementation

When constructing a component or page, always start with <code>h1</code> for the topmost level of hierarchy.

Do not rely on heading elements for styling purposes; instead use classes like <code>.callout--title</code> to style component headings.

Following these guidelines will ensure a clean document outline, which will increase both SEO performance and accessibility.

## Further Reading

<ul>
  <li>
    <a href="https://alistapart.com/article/more-meaningful-typography">More Meaningful Typography</a>
  </li>
  <li>
    <a href="https://developer.mozilla.org/en-US/docs/Web/CSS/line-height">"line-height" on MDN</a>
  </li>
</ul>

### Components Referenced

<ul>
  <li>
    <a href="/components/callout.html">Callout</a>
  </li>
  <li>
    <a href="/components/forms.html">Form</a>
  </li>
  <li>
    <a href="/components/text-input.html">Text Input</a>
  </li>
</ul>
