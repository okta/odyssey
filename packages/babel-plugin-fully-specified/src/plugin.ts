/*!
 * Copyright (c) 2021-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import type * as Babel from "@babel/core";
import * as BabelTypes from "@babel/types";
import { resolve, extname, dirname } from "path";
import { existsSync, lstatSync } from "fs";

const isNodeModule = (candidate: string): boolean => {
  if (candidate.startsWith(".") || candidate.startsWith("/")) {
    return false;
  }

  try {
    require.resolve(candidate);
    return true;
  } catch (e) {
    if (e instanceof Error) {
      const error: NodeJS.ErrnoException = e;
      if (error.code == "MODULE_NOT_FOUND") return false;
      throw error;
    }
  }
  return false;
};

const skipModule = (candidate: string) =>
  !candidate.startsWith(".") ||
  isNodeModule(candidate) ||
  [".js", ".scss"].includes(extname(candidate));

export function plugin({
  types: t,
}: {
  types: typeof BabelTypes;
}): Babel.PluginObj {
  return {
    name: "fully-specified",
    visitor: {
      // @ts-expect-error type annotations don't support multiple visitor foo|bar|baz syntactic sugar
      "ImportDeclaration|ExportNamedDeclaration|ExportAllDeclaration": (
        path: Babel.NodePath<
          | BabelTypes.ImportDeclaration
          | BabelTypes.ExportNamedDeclaration
          | BabelTypes.ExportAllDeclaration
        >,
        state: Babel.PluginPass
      ): void => {
        if (!state.filename) {
          return;
        }

        if (path.node.type === "ImportDeclaration") {
          if (path.node.importKind === "type") {
            return;
          }
        } else {
          if (path.node.exportKind === "type") {
            return;
          }
        }

        const candidate = path.node.source?.value;
        if (!candidate) {
          return;
        }
        if (skipModule(candidate)) {
          return;
        }

        const dirPath = resolve(dirname(state.filename), candidate);

        const fullySpecifiedLiteral = t.stringLiteral(
          existsSync(dirPath) && lstatSync(dirPath).isDirectory()
            ? `${candidate}${candidate.endsWith("/") ? "" : "/"}index.js`
            : `${candidate}.js`
        );

        switch (path.node.type) {
          case "ImportDeclaration":
            path.replaceWith(
              t.importDeclaration(path.node.specifiers, fullySpecifiedLiteral)
            );
            return;
          case "ExportNamedDeclaration":
            path.replaceWith(
              t.exportNamedDeclaration(
                path.node.declaration,
                path.node.specifiers,
                fullySpecifiedLiteral
              )
            );
            return;
          case "ExportAllDeclaration":
            path.replaceWith(t.exportAllDeclaration(fullySpecifiedLiteral));
            return;
          default:
            throw new Error("Invalid path node type for visitor!");
        }
      },
    },
  };
}
