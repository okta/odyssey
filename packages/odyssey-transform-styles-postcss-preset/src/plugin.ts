/*!
 * Copyright (c) 2021-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import type { PluginCreator } from "postcss";
import type { CssNanoOptions } from "cssnano";
import type { Options as AutoPrefixerOptions } from "autoprefixer";
import type PostcssModulesPlugin from "postcss-modules";
import postcss from "postcss";
import postcssModules from "postcss-modules";
import postcssLogical from "postcss-logical";
import type { LogicalOptions } from "postcss-logical";
import cssnano from "cssnano";
import autoprefixer from "autoprefixer";
import postcssTheme from "@okta/odyssey-postcss-theme";

interface PluginOptions {
  logical?: LogicalOptions;
  cssnano?: CssNanoOptions;
  autoprefixer?: AutoPrefixerOptions;
  modules?: Parameters<PostcssModulesPlugin>[number];
  theme?: false;
}

const plugin: PluginCreator<PluginOptions> = (optsArgs = {}) => {
  const opts: Required<PluginOptions> = {
    logical: {
      dir: false,
      preserve: false,
      ...optsArgs.logical,
    },
    modules: {
      generateScopedName: "ods-[hash:hex:6]",
      ...optsArgs.modules,
    },
    cssnano: {
      preset: [
        "default",
        {
          discardComments: {
            removeAll: true,
          },
        },
      ],
      ...optsArgs.cssnano,
    },
    autoprefixer: {
      ...optsArgs.autoprefixer,
    },
    theme: !!optsArgs.theme,
  };

  const plugins = [
    optsArgs.logical !== false && postcssLogical(opts.logical),
    optsArgs.modules !== false && postcssModules(opts.modules),
    optsArgs.cssnano !== false &&
      (cssnano(opts.cssnano) as PluginCreator<CssNanoOptions>),
    optsArgs.autoprefixer !== false && autoprefixer(opts.autoprefixer),
    optsArgs.theme !== false && postcssTheme(),
  ].filter(Boolean);

  return postcss(plugins);
};

plugin.postcss = true;
export { plugin as default };
