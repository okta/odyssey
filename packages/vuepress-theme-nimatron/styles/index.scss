@import '~@okta/odyssey';
@import '~prismjs/themes/prism.css';

// Doc Specific Tokens
$more-opaque: 0.6;
$med-opaque: 0.5;
$less-opaque: 0.3;

// Nimatron Specific Styles
@import 'prism-ext'; // TODO: Create prism-theme-odyssey package for this
@import 'samples';

.is-not-provisional {
  display: none;
}

.nimatron {
  display: block;
  min-height: 100vh;
  background-color: $color-primary-dark;
  background-repeat: no-repeat;

  @media only screen and (min-width: 75ch) {
    display: grid;
    grid-template-columns: minmax(auto, max-content) minmax(auto, auto);
  }
}

.nimatron--header {
  padding: $spacing-m $spacing-l $spacing-l $spacing-m;
  color: $white;
}

.nimatron--header-title {
  margin-bottom: $spacing-m;
  font-size: 0;
}

.nimatron--header-logo {
  height: $spacing-m;
  fill: $white;
}

.nimatron--nav {
  display: grid;
  grid-template-columns: minmax(auto, auto) minmax(auto, auto);
  column-gap: $spacing-m;
  text-transform: capitalize;

  @media only screen and (min-width: 75ch) {
    display: block;
  }
}

.nimatron--nav-section {
  margin: 0 0 $spacing-m;
}

.nimatron--nav-item {
  margin: 0 0 $spacing-xs;
}

.nimatron--nav-heading {
  color: $white;
  font-size: $size-body-base;
  margin: $spacing-s 0 $spacing-s;
  font-weight: 600;
}

.nimatron--nav-list {
  margin: 0 0 $spacing-m;
  list-style-type: none;
  font-weight: 400;
  font-size: 14px;

  a {
    position: relative;
    color: $white;
    text-decoration: none;

    &[target='_blank']:not([href^='mailto:'])::after {
      content: '';
      display: inline-block;
      width: 1em;
      height: 1em;
      margin-left: 0.2em;
      background-image: get-icon('external', $white);
      background-repeat: no-repeat;
      vertical-align: middle;
    }
    
    &::before {
      content: '';
      position: absolute;
      bottom: 0;
      left: 0;
      width: 100%;
      transition: border-bottom-color $base-duration $base-timing;
      border-bottom: 1px solid;
      border-bottom-color: rgba($white, $less-opaque);
    }

    &:hover {
      &::before {
        border-bottom-color: rgba($white, $more-opaque);
      }
    }
  }
}

.nimatron--main {
  display: block;
  background-color: $white;

  @media only screen and (min-width: 75ch) {
    display: grid;
    grid-template-columns: minmax(auto, ($max-line-length * 2)) minmax(auto, auto);
    min-width: 0;
    box-shadow: 0 2px $base-font-size rgba(0, 0, 0, 0.2);
  }

  h2:not([class]) {
    position: relative;

    &::before {
      content: '';
      position: absolute;
      display: none;
      top: 0;
      left: -$spacing-l;
      width: calc(100% + #{$spacing-l} + #{$spacing-l});
      border-top: 1px solid cv('gray', '200');
    }
  }
}

.nimatron--article {
  min-width: 0;
  padding: $spacing-m $spacing-l;
}

.nimatron--toc {
  display: none;
  position: relative;
  font-size: $size-body-caption;

  @media only screen and (min-width: 125ch) {
    display: block;
    border-left: 1px solid cv('gray', '200');
  }
}

.nimatron--toc-space-holder,
.nimatron--toc-fixed {
  padding: $spacing-m;
}

.nimatron--toc-space-holder {
  color: transparent;
  pointer-events: none;
}

.nimatron--toc-fixed {
  position: fixed;
}

.nimatron--toc-title {
  font-size: $size-body-base;
  margin-bottom: $spacing-s;
  font-weight: 600;
}

.nimatron--toc-list-item {
  margin-bottom: 0;
}

.nimatron--toc-list-level-2 {
  margin-bottom: $spacing-xs;
}

.nimatron--toc-list-level-3 {
  .nimatron--toc-list-child {
    display: none;
  }
}
.nimatron--toc-list-link {
  position: relative;
  color: $text-sub;
  text-decoration: none;

  &::before {
    content: '';
    position: absolute;
    bottom: 0;
    left: 0;
    width: 100%;
    transition: border-bottom-color $base-duration $base-timing;
    border-bottom: 1px solid;
    border-bottom-color: transparent;
  }

  &:hover {
    &::before {
      border-bottom-color: $color-primary-base;
    }
  }

  &[href='#undefined'] {
    display: none;
  }

  .nimatron--toc-space-holder & {
    color: transparent;

    &::before {
      display: none;
    }
  }

  .nimatron--toc-list-level-2 > & {
    color: $text-body;
    font-weight: 600;

    .nimatron--toc-space-holder & {
      color: transparent;
    }
  }
}

.nimatron--example {
  max-width: 100%;
  margin-bottom: $spacing-m;
}

.nimatron--rendered {
  padding: $spacing-m;
  border: 1px dashed cv('gray', '200');
  border-radius: $base-border-radius;

  &:not(:only-child) {
    border-bottom: 0;
  }

  &.is-example-disallowed {
    position: relative;
    border-color: $text-danger;

    &::after {
      content: "This is a Don't.";
      position: absolute;
      bottom: 0;
      right: 0;
      padding: 0 $spacing-s $spacing-s 0;
      font-size: $size-body-caption;
      font-weight: 600;
      color: cv('red', '900');
    }
  }
}

.nimatron--figure {
  margin-bottom: $spacing-m;
}

svg {
  max-width: 100%;
}

// Hexo Render Hack

h1 {
  &:empty {
    margin: 0;
  }
}

.content__nimatron-all { // stylelint-disable
  > h1,
  > h2,
  > h3,
  > h4,
  > h5,
  > h6 {
    position: relative;

    .header-anchor {
     display: none;
    }
  }
}
