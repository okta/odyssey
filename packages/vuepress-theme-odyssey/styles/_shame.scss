// Manually adjust spacing between these elements.
// Can be removed when type scale / heading margins are fixed.

h2 + .docskit-figure-anatomy {
  //margin-top: $spacing-xs;
}
