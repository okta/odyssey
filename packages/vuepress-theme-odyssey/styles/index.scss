@import '~@okta/odyssey';
@import '~@okta/prism-theme-odyssey';

// Internal
@import 'abstracts/theming';
@import 'abstracts/mixins';
@import 'abstracts/functions';
@import 'base/fonts';
@import 'components/DocsBanner';
@import 'components/DocsCard';
@import 'components/DocsCardGroup';
@import 'components/DocsLink';
@import 'components/DocsNav';
@import 'components/DocsTopbar';
@import 'components/DocsSidebar';
@import 'components/DocsPageHeader';
@import 'components/DocsPagination';
@import 'templates/DocsTemplateHome';
@import 'templates/DocsTemplateComponent';
@import 'templates/DocsTemplatePlain';
@import 'docskit/utility-classes';
@import 'docskit/Anatomy';
@import 'docskit/marker';
@import 'docskit/Visual';
@import 'docskit/Description';

// Shame file for known hacks
@import 'shame';

html {
  @include set-theme($theme-default);

  background: var(--docs-page-bg, #{$white});
  color: var(--docs-text-body, #{$text-body});
  font-family: $body-font-family;

  // Theme: Night Mode
  // @media (prefers-color-scheme: dark) {
  //   @include set-theme($theme-night);
  // }
}

a,
a:visited {
  &:focus {
    @include outline;
  }
}

.docs-layout {
  display: flex;

  .main--container {
    display: flex;
    flex: 1;
    justify-content: center;
    max-width: 100%;
    margin-bottom: $spacing-l;
    padding: $spacing-l * 2 $spacing-m 0;

    // TODO: We would like to remove this in the future
    // so that the grid width is determined by the content
    // as opposed to a max-width.
    main {
      width: 100%;
      max-width: calc((#{$max-line-length} * 2) + #{$spacing-l});
    }
  }
}

.docs-skip-content {
  position: absolute;
  z-index: 20;
  top: -3rem;
  left: $spacing-xs;
  padding: $spacing-xs $spacing-s;
  overflow: hidden;
  transition: 0.15s ease-in-out;
  background: $white;
  font-weight: 600;
  text-decoration: none;

  &:focus {
    top: $spacing-xs;
    width: auto;
    height: auto;
  }
}

.docs-banner--beta {
  flex-basis: 100%;

  @include mq(l) {
    position: sticky;
    z-index: z-index(topbar);
    top: 0;
  }
}

h1,
h2,
h3,
h4,
h5,
h6 {
  .header-anchor {
    display: none;
    position: absolute;
    width: $spacing-s-em;
    height: $spacing-s-em;
    margin-top: $spacing-xs-em /2;
    margin-left: -$spacing-m;
    transition: 0.33s opacity  ease-in-out;
    opacity: 0;
    background-image: get-icon('anchor', var(--border-color-ui, #{$border-color-ui}));
    background-repeat: no-repeat;
    background-size: contain;
    text-indent: -9999em;
    vertical-align: middle;
  }

  &:hover .header-anchor,
  &:focus-within .header-anchor {
    opacity: 1;
  }
}

// Component Samples

.is-sample-unimportant {
  opacity: 0.33;
}

.is-tab-small-sample {
  .ods-tabs--tabpanel [role='tabpanel'] {
    padding-bottom: $spacing-s;
  }
}

.is-tab-stacked-example {
  .ods-tabs [role='tabpanel'] {
    padding: $spacing-l;
  }

  .ods-tabs:first-child [role='tabpanel'] {
    padding: $spacing-xs;
  }
}

.sample-color {
  display: inline-block;
  width: 1em;
  height: 1em;
  margin: 0 auto;
  border-radius: $base-border-radius;
  vertical-align: middle;
}

@each $hue, $variants in $colors {
  @each $variant, $color in $variants {
    .sample-color--#{$hue}-#{$variant} {
      background: cv($hue, $variant);
    }
  }
}

// Design Token Examples

.sample-token {
  display: block;
  width: 1em;
  height: 1em;
  margin: 0 auto;
  border-radius: $base-border-radius;
}

// Token Examples: Color
.sample-token--color-primary-light {
  background: $color-primary-light;
}

.sample-token--color-primary-base {
  background: $color-primary-base;
}

.sample-token--color-primary-dark {
  background: $color-primary-dark;
}

.sample-token--color-danger-light {
  background: $color-danger-light;
}

.sample-token--color-danger-base {
  background: $color-danger-base;
}

.sample-token--color-danger-dark {
  background: $color-danger-dark;
}

.sample-token--color-primary-bg {
  background: $color-primary-base;
}

// Token Examples: Background
.sample-token--color-primary-bg-disabled {
  background: $color-primary-light;
}

.sample-token--color-danger-bg {
  background: $color-danger-base;
}

.sample-token--color-danger-bg-disabled {
  background: $color-danger-light;
}

// Token Examples: Outline
.sample-token--color-primary-outline {
  background: $focus-ring-primary;
}

.sample-token--color-danger-outline {
  background: $focus-ring-danger;
}

// Token Examples: Shadow
.sample-token--box-shadow-default {
  box-shadow: $box-shadow-default;
}

.sample-token--text-body {
  color: $text-body;
}

.sample-token--text-heading {
  color: $text-heading;
}

.sample-token--text-sub {
  color: $text-sub;
}

.sample-token--text-danger {
  color: $text-danger;
}

.type-body-sample--table {
  white-space: nowrap;
}

.type-title-sample--table {
  white-space: nowrap;

  .type-sample--example {
    font-weight: 600;
  }
}

.sample--emoji {
  display: inline-block;
  margin-right: $spacing-xs;
}

.sample--list {
  list-style-type: none;
}

$spacing-units: (
  'spacing-xl': $spacing-xl,
  'spacing-l': $spacing-l,
  'spacing-m': $spacing-m,
  'spacing-s': $spacing-s,
  'spacing-xs': $spacing-xs,
);

.type-sample--wrap {
  max-width: $max-line-length;
  overflow-x: scroll;
}

.type-sample--table {
  white-space: nowrap;
}

@for $i from 7 through 1 {
  $x: $i - 1;
  $n: 7 - $i;

  .type-sample:nth-of-type(#{$n}) {
    .type-sample--rem {
      $rem: ms($x);
      $rem: inspect($rem);
      $rem: str-slice($rem, 1, -4);
      $rem: str-slice($rem, 1, 4);
      $rem: $rem + 'rem';

      &::before {
        content: '#{$rem}';
      }
    }

    .type-sample--px {
      &::before {
        content: '#{round(rem-to-px(ms($x))))}';
      }
    }

    .type-sample--example {
      font-size: ms($x);
    }
  }
}

@for $i from 4 through 1 {
  $x: $i - 2;
  $n: 4 - $i;

  .type-sample-body:nth-of-type(#{$n}) {
    .type-sample-body--rem {
      $rem: ms($x);
      $rem: inspect($rem);
      $rem: str-slice($rem, 1, -4);
      $rem: str-slice($rem, 1, 4);
      $rem: $rem + 'rem';

      &::before {
        content: '#{$rem}';
      }
    }

    .type-sample-body--px {
      &::before {
        content: '#{round(rem-to-px(ms($x)))}';
      }
    }

    .type-sample-body--example {
      font-size: ms($x);
    }
  }
}

@each $name, $value in $spacing-units {
  .conversion--#{$name}-actual {
    &::before {
      content: '#{$value}';
    }
  }

  .conversion--#{$name}-px {
    &::before {
      content: '#{round(rem-to-px($value))}';
    }
  }

  .sample--type-#{$name} {
    position: relative;
    min-height: $base-font-size;
    margin-bottom: $spacing-m;
    padding: 0 0 0 $spacing-l;
    font-size: $value;

    &::before {
      content: 'ms(#{str-slice($name, 4)})';
      position: absolute;
      top: 0;
      left: 0;
      font-size: $size-body-caption;
    }
  }

  .sample--spacing-#{$name} {
    position: relative;
    width: $value;
    height: $value;
    background: $color-danger-light;
  }
}

.sample--tip {
  display: inline-block;
  margin: 0 $spacing-xs;
}

.help-icon {
  width: $spacing-l;
  height: $spacing-l;
  margin-bottom: $spacing-m;
}

.is-sample-absolute {
  position: absolute;
}

.is-sample-color-green {
  color: cv('green', '500');
}
