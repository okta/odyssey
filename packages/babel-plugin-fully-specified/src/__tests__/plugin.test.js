/*!
 * Copyright (c) 2021-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { transformSync } from "@babel/core";
import plugin from "../../dist";

const importStatements = `
import { level } from '.'
import { levelIndex } from './'
import { oneBackLevel } from '..'
import { oneBackLevelIndex } from '../'
import { twoBackLevel } from '../..'
import { twoBackLevelIndex } from '../../'
import { somethingBack } from '../lib/something'
import { somethingBackTest } from '../lib/something.test'
import { export1 , export2 as alias2 } from './lib/something'
import { export1Test , export2 as alias2Test } from './lib/something.test'
import { something } from './lib/something'
import { somethingTest } from './lib/something.test'
import { something as other } from './lib/something'
import { something as otherTest } from './lib/something.test'
import anotherImport from './lib/something'
import anotherImportTest from './lib/something.test'
import another, { otherImport } from './lib/something'
import anotherTest, { otherImportTest } from './lib/something.test'
import * as Something from './lib/something'
import * as SomethingTest from './lib/something.test'
`;

const exportStatements = `
export { level } from '.'
export { levelIndex } from './'
export { oneBackLevel } from '..'
export { oneBackLevelIndex } from '../'
export { twoBackLevel } from '../..'
export { twoBackLevelIndex } from '../../'
export { somethingBack } from '../lib/something'
export { somethingBackTest } from '../lib/something.test'
export { something } from './lib/something'
export { somethingTest } from './lib/something.test'
export { something as another } from './lib/something'
export { somethingTest as anotherTest } from './lib/something.test'
export * as anotherModule from './lib/something'
export * as anotherModuleTest from './lib/something.test'
export * from './lib/something'
export * from './lib/something.test'
`;

const typeOnlyExports = `
export type { NamedType } from './lib/something'
export type { NamedTypeTest } from './lib/something.test'
`;

const typeOnlyImports = `
import type DefaultType from './lib/something'
import type DefaultTypeTest from './lib/something.test'
import type { NamedType } from './lib/something'
import type { NamedTypeTest } from './lib/something.test'
import type * as AllTypes from './lib/something'
import type * as AllTypesTest from './lib/something.test'
`;

const styleModuleImports = `
import stylesWithMultipleExtensions from './styles.module.scss'
import styles from './styles.scss'
`;

const bareModuleSpecifiers = `
import babel from '@babel/core'
import { transform } from '@babel/core'

export * from '@babel/core'
export { transform } from '@babel/core'
`;

describe("plugin", () => {
  test.each`
    type                | statements
    ${"module imports"} | ${importStatements}
    ${"module exports"} | ${exportStatements}
  `("should fully qualify the $type specifiers", ({ statements }) => {
    const result = transformSync(statements, {
      plugins: ["@babel/plugin-syntax-typescript", plugin],
      filename: "foo",
      configFile: false,
    });

    expect(result?.code).toMatchSnapshot();
  });

  test.each`
    type                        | statements
    ${"type-only imports"}      | ${typeOnlyImports}
    ${"type-only exports"}      | ${typeOnlyExports}
    ${"style module imports"}   | ${styleModuleImports}
    ${"bare module specifiers"} | ${bareModuleSpecifiers}
  `("should skip $type", ({ statements }) => {
    const result = transformSync(statements, {
      plugins: ["@babel/plugin-syntax-typescript", plugin],
      filename: "foo",
      configFile: false,
    });

    expect(result?.code).toMatchSnapshot();
  });
});
