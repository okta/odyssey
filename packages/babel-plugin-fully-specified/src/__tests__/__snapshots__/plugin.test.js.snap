// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin should fully qualify the module exports specifiers 1`] = `
"export { level } from "./index.js";
export { levelIndex } from "./index.js";
export { oneBackLevel } from "../index.js";
export { oneBackLevelIndex } from "../index.js";
export { twoBackLevel } from "../../index.js";
export { twoBackLevelIndex } from "../../index.js";
export { somethingBack } from "../lib/something.js";
export { somethingBackTest } from "../lib/something.test.js";
export { something } from "./lib/something.js";
export { somethingTest } from "./lib/something.test.js";
export { something as another } from "./lib/something.js";
export { somethingTest as anotherTest } from "./lib/something.test.js";
export * as anotherModule from "./lib/something.js";
export * as anotherModuleTest from "./lib/something.test.js";
export * from "./lib/something.js";
export * from "./lib/something.test.js";"
`;

exports[`plugin should fully qualify the module imports specifiers 1`] = `
"import { level } from "./index.js";
import { levelIndex } from "./index.js";
import { oneBackLevel } from "../index.js";
import { oneBackLevelIndex } from "../index.js";
import { twoBackLevel } from "../../index.js";
import { twoBackLevelIndex } from "../../index.js";
import { somethingBack } from "../lib/something.js";
import { somethingBackTest } from "../lib/something.test.js";
import { export1, export2 as alias2 } from "./lib/something.js";
import { export1Test, export2 as alias2Test } from "./lib/something.test.js";
import { something } from "./lib/something.js";
import { somethingTest } from "./lib/something.test.js";
import { something as other } from "./lib/something.js";
import { something as otherTest } from "./lib/something.test.js";
import anotherImport from "./lib/something.js";
import anotherImportTest from "./lib/something.test.js";
import another, { otherImport } from "./lib/something.js";
import anotherTest, { otherImportTest } from "./lib/something.test.js";
import * as Something from "./lib/something.js";
import * as SomethingTest from "./lib/something.test.js";"
`;

exports[`plugin should skip bare module specifiers 1`] = `
"import babel from '@babel/core';
import { transform } from '@babel/core';
export * from '@babel/core';
export { transform } from '@babel/core';"
`;

exports[`plugin should skip style module imports 1`] = `
"import stylesWithMultipleExtensions from './styles.module.scss';
import styles from './styles.scss';"
`;

exports[`plugin should skip type-only exports 1`] = `
"export type { NamedType } from './lib/something';
export type { NamedTypeTest } from './lib/something.test';"
`;

exports[`plugin should skip type-only imports 1`] = `
"import type DefaultType from './lib/something';
import type DefaultTypeTest from './lib/something.test';
import type { NamedType } from './lib/something';
import type { NamedTypeTest } from './lib/something.test';
import type * as AllTypes from './lib/something';
import type * as AllTypesTest from './lib/something.test';"
`;
